### ISSUE TITLE
ENH: integrate the interface to get prediction/forecast results in UnobservedComponents

### ISSUE THREAD

**Issue by tatamiya** [statsmodels/statsmodels#7344]
#### Is your feature request related to a problem? Please describe

Sometimes I have difficulty in getting prediction/forecast results of UnobservedComponents.

Because I have to use different methods and properties to get the results depending on the situation,
I often get confused which one to use. Moreover, the code tends to get long and complicated.

In the current implementation, there are two ways to get results:

- UnobservedComponentsResults.get_prediction()
    - one-step-ahead prediction and forecast (sum of all components)
- level, trend, seasonal, cycle properties of UnobservedComponentResults
    - filtered or smoothed state and its covariance for each component

Moreover, in some cases no interface provides the results we'd like to get 
and we have to look for other methods and properties.
For example, 

- out-of-sample forecast for each component
- prediction/forecast results of mixed components (especially variance)

Here is an example of local level + seasonality model (see the details below):
<details>

```python
import statsmodels.api as sm

nile = sm.datasets.get_rdataset("Nile")

model = sm.tsa.UnobservedComponents(nile.data['value'], level='local level', seasonal=10)
results = model.fit()
forecast = results.get_forecast(steps=4)
```

In order to get the out-of-sample forecast for level component, we have to access PredictionResults.prediction_results.results:

```python
# forecast of level component
forecast.prediction_results.results.predicted_state[0]
# forecast variance of level component
forecast.prediction_results.results.predicted_state_cov[0][0]
```

In order to get smoothed variance of level + seasonality (without irregularity component), We have to find covariance matrix.

```python
# smoothed states of level + seasonal components
results.level['smoothed'] + results.seasonal['smoothed']

# smoothed variance of level + seasonal component
results.smoothed_state_cov[:2, :2].sum(0).sum(0)
```

</details>
On the other hand, when you use KFAS of R, you can get the same results only with `predict` function.

#### Describe the solution you'd like

I'd like to make a unified interface by extending `get_prediction` method of UnobservedComponentResults.
It would be simple to get every possible pattern of prediction/forecast only with this method.

My idea is that we need three new option arguments:

- state: str
Select which state you get, `one-step-ahead`, `filtered`, or `smoothed`.
If  the selected state does not exist, use `one-step-ahead` (e.g. out-of-sample forecast).

- component: str or list of str
Select which component you get `level`, `trend`, `seasonal`, `cycle`, `exog`, or `all`.
If the input is list of component names (such as `['level', 'exog']`), return sum of the components.

- observed: bool
If true, add observation noise covariance to the result.
(In KFAS of R, it corresponds to `interval={'confidence' / 'prediction'}` in `predict` function)

**Comment by ChadFulton** [comment]
Yes, this is a good idea, and could be added to all state space models. I don't think it should be too hard.

**Comment by bashtage** [comment]
> * Select which state you get, `one-step-ahead`, `filtered`, or `smoothed`.
>   If  the selected state does not exist, use `one-step-ahead` (e.g. out-of-sample forecast).

I think in cases where something requested doesn't exist an exception should be raised.  Providing something other than what is requested seems to violate the Pythonic principle that explicit is better than implicit.


> * observed: bool
>   If true, add observation noise covariance to the result.
>   (In KFAS of R, it corresponds to `interval={'confidence' / 'prediction'}` in `predict` function)

Observed doesn't sound like a great name.  noise, covariance or add_covariance all seem clearer with the intent.

**Comment by tatamiya** [comment]
Thank you!

>could be added to all state space models.

I'll also check other statespace classes (such as ExponentialSmoothing and SARIMAX).

>I think in cases where something requested doesn't exist an exception should be raised. Providing something other than what is requested seems to violate the Pythonic principle that explicit is better than implicit.

I agree.

> Observed doesn't sound like a great name.

Then, `add_observation_noise` might be better.
I have another idea that the observation noise should be treated as one of the components (`irregularity`).
